src/Controller/WebSocketController.ts

```diff
@@ -0,0 +1,104 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { getVersion } from '@manuscripts/transform'
+import { IncomingMessage } from 'http'
+import { Duplex } from 'stream'
+import { ErrorEvent, MessageEvent, WebSocket, WebSocketServer } from 'ws'
+
+import { DIContainer } from '../DIContainer/DIContainer'
+import { AuthorityService } from '../DomainServices/AuthorityService'
+import { DocumentPermission, DocumentService } from '../DomainServices/DocumentService'
+import { SocketsService } from '../DomainServices/SocketsService'
+import { getManuscriptIDFromRequest } from '../util'
+import { log } from '../Utilities/Logger'
+
+export class WebSocketController {
+  private socketsService: SocketsService
+  private authorityService: AuthorityService
+  private documentService: DocumentService
+  constructor(private readonly wss: WebSocketServer) {
+    this.socketsService = DIContainer.sharedContainer.socketsService
+    this.authorityService = DIContainer.sharedContainer.authorityService
+    this.documentService = DIContainer.sharedContainer.documentService
+  }
+
+  public handleUpgrade = async (request: IncomingMessage, socket: Duplex, head: Buffer) => {
+    this.wss.handleUpgrade(request, socket, head, async (ws) =>
+      this.setupConnection(socket, ws, request)
+    )
+  }
+
+  private setupConnection = async (stream: Duplex, ws: WebSocket, request: IncomingMessage) => {
+    const manuscriptID = getManuscriptIDFromRequest(request)
+    this.wss.emit('connection', ws, request)
+    this.attachListeners(stream, ws, manuscriptID)
+    log.info(`Connection established for ${manuscriptID}`)
+  }
+
+  private attachListeners = (stream: Duplex, ws: WebSocket, manuscriptID: string) => {
+    ws.onerror = (error) => this.onError(ws, error)
+    ws.onclose = () => this.onClose(ws, manuscriptID)
+    ws.onmessage = (message) => this.onMessage(stream, ws, message)
+  }
+
+  private onError = (ws: WebSocket, error: ErrorEvent) => {
+    log.error(`WebSocket error: ${error}`)
+    this.socketsService.closeSocket(ws)
+  }
+  private onClose = (ws: WebSocket, manuscriptID: string) => {
+    log.info(`Connection closed for ${manuscriptID}`)
+    this.socketsService.closeSocket(ws, manuscriptID)
+  }
+
+  private onMessage = async (stream: Duplex, ws: WebSocket, message: MessageEvent) => {
+    const { manuscriptID, projectID, authToken } = this.parseMessage(message)
+    if (!manuscriptID || !projectID || !authToken) {
+      return this.handleInvalidData(stream, ws, manuscriptID)
+    }
+    try {
+      await this.documentService.validateTokenAccess(authToken, projectID, DocumentPermission.READ)
+      this.socketsService.setClient(manuscriptID, ws)
+      await this.sendHistoryToSocket(manuscriptID, ws)
+    } catch (error) {
+      log.error(`Error handling message: ${error}`)
+      this.handleInvalidData(stream, ws, manuscriptID)
+    }
+  }
```

Comment:
So the current flow expects the FE to send these information on startup to open the connection. But we technically can get this information and validate from the request url when we receive the request, which do you prefer?  @mnatsheh